File: backend/src/schemas/character.py
"""Character Pydantic schemas."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class AbilityScores(BaseModel):
    """D&D 5e ability scores."""

    strength: int = Field(..., ge=1, le=20, description="Strength score (1-20)")
    dexterity: int = Field(..., ge=1, le=20, description="Dexterity score (1-20)")
    constitution: int = Field(..., ge=1, le=20, description="Constitution score (1-20)")
    intelligence: int = Field(..., ge=1, le=20, description="Intelligence score (1-20)")
    wisdom: int = Field(..., ge=1, le=20, description="Wisdom score (1-20)")
    charisma: int = Field(..., ge=1, le=20, description="Charisma score (1-20)")


class CharacterCreate(BaseModel):
    """Schema for creating a new character."""

    name: str = Field(..., min_length=1, max_length=100, description="Character name")
    character_class: str = Field(
        ...,
        alias="class",
        description="D&D 5e class (barbarian, bard, cleric, druid, fighter, monk, paladin, ranger, rogue, sorcerer, warlock, wizard)",
    )
    race: str = Field(
        ...,
        description="D&D 5e race (dragonborn, dwarf, elf, gnome, half-elf, halfling, half-orc, human, tiefling)",
    )
    ability_scores: AbilityScores = Field(..., description="Character ability scores")
    level: int = Field(default=1, ge=1, le=20, description="Character level (1-20)")
    backstory: str | None = Field(
        default=None, max_length=5000, description="Character backstory"
    )

    model_config = ConfigDict(populate_by_name=True)


class CharacterUpdate(BaseModel):
    """Schema for updating a character."""

    name: str | None = Field(default=None, min_length=1, max_length=100)
    backstory: str | None = Field(default=None, max_length=5000)
    level: int | None = Field(default=None, ge=1, le=20)

    model_config = ConfigDict(populate_by_name=True)


class CharacterResponse(BaseModel):
    """Schema for character response."""

    id: UUID
    name: str
    character_class: str = Field(..., serialization_alias="class")
    race: str
    level: int
    ability_scores: AbilityScores
    backstory: str | None
    created_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
    )


class CharacterListResponse(BaseModel):
    """Schema for list of characters."""

    characters: list[CharacterResponse]
    total: int

File: ios/lib/features/character/models/character.dart
import 'ability_scores.dart';

/// Модель персонажа D&D 5e
class Character {
  final String id;
  final String name;
  final String characterClass;
  final String race;
  final AbilityScores abilityScores;
  final DateTime createdAt;
  final int level;
  final String? backstory;
  final DateTime? updatedAt;

  const Character({
    required this.id,
    required this.name,
    required this.characterClass,
    required this.race,
    required this.abilityScores,
    required this.createdAt,
    this.level = 1,
    this.backstory,
    this.updatedAt,
  });

  factory Character.fromJson(Map<String, dynamic> json) {
    return Character(
      id: json['id'] as String,
      name: json['name'] as String,
      characterClass: json['class'] as String,
      race: json['race'] as String,
      abilityScores: AbilityScores.fromJson(json['ability_scores'] as Map<String, dynamic>),
      createdAt: DateTime.parse(json['created_at'] as String),
      level: (json['level'] as num?)?.toInt() ?? 1,
      backstory: json['backstory'] as String?,
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at'] as String) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'class': characterClass,
      'race': race,
      'ability_scores': abilityScores.toJson(),
      'created_at': createdAt.toIso8601String(),
      'level': level,
      if (backstory != null) 'backstory': backstory,
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
    };
  }

  /// Бонус мастерства по уровню (D&D 5e)
  int get proficiencyBonus => ((level - 1) ~/ 4) + 2;

  /// Максимальное здоровье (упрощённый расчёт)
  int get maxHitPoints {
    final hitDie = _getHitDie(characterClass);
    final conMod = abilityScores.constitutionModifier;
    // Первый уровень: максимум кубика + con mod
    // Остальные уровни: среднее + con mod
    if (level == 1) {
      return hitDie + conMod;
    }
    final averagePerLevel = (hitDie ~/ 2) + 1;
    return hitDie + conMod + (level - 1) * (averagePerLevel + conMod);
  }

  int _getHitDie(String className) => switch (className.toLowerCase()) {
      'barbarian' || 'варвар' => 12,
      'fighter' || 'воин' || 'paladin' || 'паладин' || 'ranger' || 'следопыт' =>
        10,
      'bard' ||
      'бард' ||
      'cleric' ||
      'жрец' ||
      'druid' ||
      'друид' ||
      'monk' ||
      'монах' ||
      'rogue' ||
      'плут' ||
      'warlock' ||
      'колдун' =>
        8,
      'sorcerer' || 'чародей' || 'wizard' || 'волшебник' => 6,
      _ => 8,
    };
}

/// Запрос на создание персонажа
class CreateCharacterRequest {
  final String name;
  final String characterClass;
  final String race;
  final AbilityScores abilityScores;
  final String? backstory;

  const CreateCharacterRequest({
    required this.name,
    required this.characterClass,
    required this.race,
    required this.abilityScores,
    this.backstory,
  });

  factory CreateCharacterRequest.fromJson(Map<String, dynamic> json) {
    return CreateCharacterRequest(
      name: json['name'] as String,
      characterClass: json['character_class'] as String,
      race: json['race'] as String,
      abilityScores: AbilityScores.fromJson(json['ability_scores'] as Map<String, dynamic>),
      backstory: json['backstory'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'character_class': characterClass,
      'race': race,
      'ability_scores': abilityScores.toJson(),
      if (backstory != null) 'backstory': backstory,
    };
  }
}

/// Запрос на обновление персонажа
class UpdateCharacterRequest {
  final String? name;
  final String? backstory;

  const UpdateCharacterRequest({
    this.name,
    this.backstory,
  });

  factory UpdateCharacterRequest.fromJson(Map<String, dynamic> json) {
    return UpdateCharacterRequest(
      name: json['name'] as String?,
      backstory: json['backstory'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (name != null) 'name': name,
      if (backstory != null) 'backstory': backstory,
    };
  }
}
